### PR TITLE
Cache and validate devfile samples

### PR DESCRIPTION
**Pull request by johnmcollier**
**What does does this PR do / why we need it**:
- Updates the registry build to download and cache the devfile samples in a registry
    - Downloads the git repo to a zip, and also extracts the devfile.yaml and sample icon (if present)
    - Saves them to `samples/<sample-name>` under the registry build output folder
- Updates the index generation to also validate devfile samples _if_ the samples have been cached 
    - I chose the index generator to do the validation since it was already validating the devfile stacks there, and we don't have access to the devfile library 
- Changes the folder where the registry resources are stored on the container from `/stacks` and `/index.json` to `/registry/stacks`, `/registry/samples` and `/registry/index.json`.

**Which issue(s) this PR fixes**:

Fixes https://github.com/devfile/api/issues/480

**PR acceptance criteria**:

- [x] Test:
    - Updated the the test registry data for the index generator tests to cover this scenario.

- [x] Documentation
    - Not applicable.

### CODE REVIEW

File: index/server/Dockerfile [hunk 1]
```diff
@@ -36,8 +36,9 @@ RUN chmod g+rwx /var/run/nginx.pid && \
 USER www-data
 
 # Set env vars for the locations of the devfile stacks and index.json
-ENV DEVFILE_STACKS /stacks
-ENV DEVFILE_INDEX /index.json
+ENV DEVFILE_STACKS /registry/stacks
+ENV DEVFILE_SAMPLES /registry/stacks
```
Comment (GeekArthur) [inline index/server/Dockerfile:40]: Should this be `/registry/samples`?

**Review by GeekArthur — CHANGES_REQUESTED** [review]

File: build-tools/cache_samples.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+
+# This script downloads and caches the devfile samples in the registry
+# This is only called if extraDevfilesEntries.yaml exists and has entries for devfile samples
+# The downloaded samples are cached under /samples in the devfile registry container
```
[the remark this answers is not in the document]
Comment (johnmcollier) [reply build-tools/cache_samples.sh:5]: Yup, you're right. Initially I was using /samples, but had to change to `/registry/samples` due to how the docker build works.

File: build-tools/Dockerfile [hunk 3]
```diff
@@ -6,7 +6,6 @@
 #
 # SPDX-License-Identifier: EPL-2.0
 #
-FROM quay.io/devfile/devfile-index-base:next
+FROM quay.io/openshift-pipeline/golang:1.15-alpine AS builder
```
[the remark this answers is not in the document]
Comment (johnmcollier) [reply build-tools/Dockerfile:9]: 🤦 I have no idea how this happened, but you're right.

File: index/server/Dockerfile (hunk 1 above)
Comment (johnmcollier) [reply index/server/Dockerfile:40]: Good catch, that was a bad rebase on my end 🤦

**Comment by johnmcollier** [comment]
@GeekArthur Review comments addressed, thanks!

**Review by GeekArthur — APPROVED** [review]

**Comment by maysunfaisal** [comment]
@johnmcollier we can do this later but the only concern i have with samples is what happens when devfile.yaml is not at the roof of a sample repo and this is somethinf ODC handles in their UI

**Comment by johnmcollier** [comment]
@maysunfaisal Yeah, let's handle this later. I didn't want to search the entire sample subtree for devfiles, so I left it out for now. Maybe we need to revisit storing the devfile.yaml location in the sample metadata (like we do with the icon)?

**Comment by maysunfaisal** [comment]
@johnmcollier yeah, the simplest and convenient way would be to store the devfile location relative to the root in the index metadata as the sample author would have this info and we can just read it from there. 🤔

**Pull request merged**